components/range_cache_memory_engine/src/background.rs

```diff
@@ -956,34 +970,90 @@ impl Runnable for BackgroundRunner {
+                            let now = match block_on_timeout(pd_client.get_tso(), tso_timeout) {
+                                Ok(Ok(ts)) => ts,
+                                err => {
+                                    error!(
+                                        "get timestamp failed, skip gc loaded range";
+                                        "timeout_duration" => ?tso_timeout,
+                                        "error" => ?err,
+                                    );
+                                    // Get timestamp fail so don't do gc.
+                                    return Some(0);
+                                }
+                            };
+
+                            let safe_point = (|| {
+                                fail::fail_point!("in_memory_engine_safe_point_in_loading", |t| {
+                                    t.unwrap().parse::<u64>().unwrap()
+                                });
+
+                                let safe_point = now
+                                    .physical()
+                                    .saturating_sub(gc_interval.as_millis() as u64);
+                                TimeStamp::compose(safe_point, 0).into_inner()
+                            })();
+
+                            let write_cf_handle = skiplist_engine.cf_handle(CF_WRITE);
+                            let default_cf_handle = skiplist_engine.cf_handle(CF_DEFAULT);
+                            let mut filter = Filter::new(
+                                safe_point,
+                                u64::MAX,
+                                default_cf_handle,
+                                write_cf_handle.clone(),
+                            );
+
+                            let mut iter = write_cf_handle.iterator();
+                            let guard = &epoch::pin();
+                            let (start_key, end_key) = encode_key_for_boundary_with_mvcc(&range);
+                            iter.seek(&start_key, guard);
+                            while iter.valid() && iter.key() < &end_key {
+                                let k = iter.key();
+                                let v = iter.value();
+                                if let Err(e) = filter.filter(k.as_bytes(), v.as_bytes()) {
+                                    warn!(
+                                        "Something Wrong in memory engine GC";
+                                        "error" => ?e,
+                                    );
+                                }
+                                iter.next(guard);
+                            }
+
+                            Some(safe_point)
                         };
 
                         let start = Instant::now();
-                        if !snapshot_load() {
+                        if let Some(safe_point) = snapshot_load() {
+                            if core.on_snapshot_load_finished(
```

Comment:
```suggestion
                        if let Some(safe_point) = snapshot_load()
                            && core.on_snapshot_load_finished(
```

Reply:
why this?

Reply:
Why not directly use the extracted `safe_point` from the `Some(safe_point)` to do the later judgement ?

Reply:
I still do not understand what you mean, You seem to merge two if but there's a else for the second if. By the way, in on_snapshot_load_finished, it still can fail.